controllers/pvc-mutator/encryption/keys/manager.go

```diff
@@ -0,0 +1,173 @@
+package keys
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/storageos/api-manager/internal/pkg/crypto"
+)
+
+var (
+	// ErrNoKeyInSecret is returned is a secret was expected to contain an
+	// encryption key, but didn't.
+	ErrNoKeyInSecret = errors.New("secret does not contain encryption key")
+)
+
+// KeyManager generates, stores and removes encryption keys.
+type KeyManager struct {
+	client client.Client
+}
+
+// New creates a new KeyManager that is responsible for generationg and storing
+// volume encryption keys.
+func New(client client.Client) *KeyManager {
+	return &KeyManager{client: client}
+}
+
+// Ensure that a secret exists at volKeyRef, creating it with valid keys if
+// needed.  If a secret already exists, it does not verify validity.
+func (m *KeyManager) Ensure(ctx context.Context, nsKeyRef client.ObjectKey, volKeyRef client.ObjectKey, nsSecretLabels map[string]string, volSecretLabels map[string]string) error {
+	// Return immediately if the volume key already exists, or an error if
+	// another error (e.g. permission denied).
+	if err := m.client.Get(ctx, volKeyRef, &corev1.Secret{}); err == nil || !apierrors.IsNotFound(err) {
```

Comment:
Returning err if err is nil. Doesn't look right. :thinking:

Reply:
Just saw https://github.com/storageos/api-manager/pull/45#discussion_r614680821 . idk if it makes it hard to read.

Reply:
Agreed, putting it back.